reconcile/slack_usergroups.py

```diff
@@ -161,80 +172,50 @@ def get_slack_usernames_from_pagerduty(pagerduties, users, usergroup):
-                                           github_owners_aliases, users,
-                                           usergroup):
-    return get_slack_usernames_from_owners(
-        github_owners, github_owners_aliases, users, usergroup,
-        'github_username', missing_user_log_method=logging.debug)
+def get_slack_usernames_from_owners(owners_from_repo, users, usergroup):
+    if owners_from_repo is None:
+        return []
 
+    all_slack_usernames = []
 
-def get_slack_usernames_from_gitlab_owners(gitlab_owners,
-                                           gitlab_owners_aliases, users,
-                                           usergroup):
-    return get_slack_usernames_from_owners(
-        gitlab_owners, gitlab_owners_aliases, users, usergroup,
-        'org_username', ssl_verify=False)
+    for url in owners_from_repo:
+        repo_cli = RepoCli(url)
 
+        if isinstance(repo_cli, GitLabApi):
+            user_key = 'org_username'
+            missing_user_log_method = logging.warning
+        elif isinstance(repo_cli, GithubApi):
+            user_key = 'github_username'
+            missing_user_log_method = logging.debug
+        else:
+            raise TypeError(f'{type(repo_cli)} not supported')
 
-def get_slack_usernames_from_owners(owners_raw_urls, owners_aliases_raw_url,
-                                    users, usergroup, user_key,
-                                    ssl_verify=True,
-                                    missing_user_log_method=logging.warning):
-    all_slack_usernames = []
-    all_username_keys = [u[user_key] for u in users]
-
-    owners_aliases = {}
-    if owners_aliases_raw_url:
-        r = get_raw_owners_content(owners_aliases_raw_url,
-                                   ssl_verify=ssl_verify)
-        try:
-            content = anymarkup.parse(r.content, force_types=None)
-            owners_aliases = content['aliases']
-        except (anymarkup.AnyMarkupError, KeyError):
-            msg = "Could not parse data. Skipping owners_aliases file: {}"
-            logging.warning(msg.format(owners_aliases_raw_url))
-
-    for owners_raw_url in owners_raw_urls or []:
-        r = get_raw_owners_content(owners_raw_url, ssl_verify=ssl_verify)
-        try:
-            content = anymarkup.parse(r.content, force_types=None)
-            owners = set()
-            for value in content.values():
-                for user in value:
-                    if user in owners_aliases:
-                        for alias_user in owners_aliases[user]:
-                            owners.add(alias_user)
-                    else:
-                        owners.add(user)
-        except (anymarkup.AnyMarkupError, KeyError):
-            msg = "Could not parse data. Skipping owners file: {}"
-            logging.warning(msg.format(owners_raw_url))
-            continue
+        repo_owners = RepoOwners(git_cli=repo_cli)
 
-        if not owners:
+        owners = repo_owners.get_root_owners()
```

Comment:
why only root owners? The current integration allows for multiple owners files (although I don't know if somebody has used it)

Reply:
To keep the current behavior. No one is using multiple OWNERS in the same repo. Also, all the cases are using the root OWNERS file only.

Reply:
Current implementation:
```
venv) ~/git/qontract-reconcile (slack_owners)$ qontract-reconcile --config config.toml --dry-run slack-usergroups
[2020-04-14 23:25:51] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'gcp-project-operator', 'lseelye']
[2020-04-14 23:25:51] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'gcp-project-operator', 'jaharrin']
[2020-04-14 23:25:51] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'gcp-project-operator', 'rbastos']
[2020-04-14 23:25:51] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'gcp-project-operator', 'cblecker']
[2020-04-14 23:25:51] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'gcp-project-operator', 'jburrell']
[2020-04-14 23:25:51] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'gcp-project-operator', 'cbrumm']
```
(The additions make sense. They are are due to the `sre-platform-leads` alias here: https://github.com/openshift/gcp-project-operator/blob/master/OWNERS)

Alternative implementation, using all the `OWNERS` files in each repo (`RepoOwners().get_owners()`):
```
(venv) ~/git/qontract-reconcile (slack_owners *)$ qontract-reconcile --config config.toml --dry-run slack-usergroups
[2020-04-14 23:41:39] [WARNING] [slack_usergroups.py:get_slack_usernames_from_owners:212] - [managed-tenants] org_username not found in app-interface: ['pbrookes', 'cbrookes', 'pbrookes', 'cbrookes']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'gcp-project-operator', 'eparis']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'gcp-project-operator', 'decarr']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'gcp-project-operator', 'mturansk']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'gcp-project-operator', 'lseelye']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'gcp-project-operator', 'jaharrin']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'gcp-project-operator', 'rbastos']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'gcp-project-operator', 'cblecker']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'gcp-project-operator', 'jburrell']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'gcp-project-operator', 'cbrumm']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'rbac-permissions-operator', 'eparis']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'rbac-permissions-operator', 'decarr']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'rbac-permissions-operator', 'mturansk']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'managed-tenants', 'amoran']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'managed-tenants', 'mfreer']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'managed-tenants', 'pcremin']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'managed-tenants', 'crarobin']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'managed-tenants', 'sclarkso']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'managed-tenants', 'raananda']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'managed-tenants', 'rhopp']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'managed-tenants', 'dfestal']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'managed-tenants', 'mciccone']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'managed-tenants', 'flacatus']
[2020-04-14 23:41:43] [INFO] [slack_usergroups.py:print_diff:301] - ['add_user_to_usergroup', 'coreos', 'managed-tenants', 'jcueto']
```
(The extra additions here are due to the extra OWNERS files in the respective repositories)

WDYT?

Reply:
I have doubts. For one part, the OWNERS file spec allows for multiple owners and a single alias in the root so it would make sense to honor it. But since this is more a commodity integration which is tangentially related to the OWNERS spec and the fact is that everybody is using root, it would be bad and it would look bad to change the behaviour. I would leave the default behaviour to get only the root. If we wanted to cover all the possibilities we could add a flag in the schema, something like `recursive` that defaults to `false`, but that would mean changing the `ownersFromRepo` array of url strings to an array of objects. I like it, but I wonder if it's overdoing it 😅

Reply:
Ok, let's start with "keep the behavior" knowing that we can easily extend that.

Reply:
I would mention it in app-interface's README

Reply:
ok, done :)